app/master/build.py

```diff
@@ -309,6 +289,35 @@ def _on_enter_error_state(self, event):
         self._error_message = getattr(event, 'error_msg', default_error_msg)
         self._logger.warning('Build {} failed: {}', self.build_id(), self._error_message)
 
+    def _on_enter_preparing_state(self, event):
+        """
+        Atomize the build. This method is triggered by a state machine transition to the PREPARING state.
+        :param event: The Fysom event object
+        """
+        self._logger.info('Fetching project for build {}.', self._build_id)
+        self.project_type.fetch_project()
+        self._logger.info('Successfully fetched project for build {}.', self._build_id)
+
+        job_config = self.project_type.job_config()
+        if job_config is None:
+            raise RuntimeError('Build failed while trying to parse clusterrunner.yaml.')
+
+        subjob_calculator = getattr(event, 'subjob_calculator', None)
+        if subjob_calculator is None:
+            raise RuntimeError('Build failed as subjob_calculator is not set for this build {}.', self._build_id)
+        subjobs = subjob_calculator.compute_subjobs_for_build(self._build_id, job_config, self.project_type)
```

Comment:
When the build is canceled during atomizing, an exception `AtomizerError` is raised which causes stack trace in logs. I am still working on avoiding the stack trace when the build is canceled.
I tried catching and ignoring `AtomizerError` if the build is canceled (see below code), but that leaves the build in `PREPARING` state.

```python
        try:
            subjobs = subjob_calculator.compute_subjobs_for_build(self._build_id, job_config, self.project_type)
        except AtomizerError:
            if self.is_canceled:
                return
            else:
                raise
```